src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -355,14 +361,13 @@ private async Task OnPartialValidationCompletedCallbackAsync(ValidationContext v
             {
                 var peersToBan = new List<INetworkPeer>();
 
-                // TODO ACTIVATION implement following (used to be in consensus loop)
-                //if (validationContext.Error == ConsensusErrors.BadWitnessNonceSize)
-                //{
-                //    this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
-                //    this.connectionManager.AddDiscoveredNodesRequirement(NetworkPeerServices.NODE_WITNESS);
-                //    this.Puller.RequestOptions(TransactionOptions.Witness);
-                //    return;
-                //}
+                if (validationContext.Error == ConsensusErrors.BadWitnessNonceSize)
+                {
+                    this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                    this.connectionManager.AddDiscoveredNodesRequirement(NetworkPeerServices.NODE_WITNESS);
+                    this.blockPuller.RequestWitnessOptions();
```

Comment:
I suggest we do the same as with the connection manager and pass the network service enum as a method parameter, and this can be set in the Validation context this will make the low level oblivious to any witness changes.
Then just before we throw `ConsensusErrors.BadWitnessNonceSize` you can set the enum.

Reply:
Ok. Thought that would be needed. Will fix.

Reply:
fixed

Reply:
This is still only half the fix.
`NetworkPeerServices` can be set as a nullable type on context

Then we do 
```
if(validationContext.RequiredNetworkService != null)
   bla..
```

Reply:
I'm not sure if we want to pass services using context. 
Data in the context should be input data for the rules. So here in order to allow a rule to provide output data we will change the context for every single rule, when instead we can handle particular error like we are doing now.

Reply:
Well what I want to try to avoid is put witness related in CM so if we keep in ValidationContext the such a property (similar to how we keep the node banning flag) should be ok no?

Reply:
Fixed: https://github.com/stratisproject/StratisBitcoinFullNode/pull/2043/commits/5d6499da16f0e4e00e53d92df71e52a033b9c5bd